### PR TITLE
only ping once per page load per dbh thread

### PR DESCRIPTION
**Pull request by kkmuffme**
https://github.com/stuttter/ludicrousdb/issues/63
* add non-persistent group to store thread id connection during one page load
* removes heartbeat check

### CODE REVIEW

**Comment by JJJ** [comment]
#85 and #86 were merged, resulting in this one being in conflict.

I'm not entirely sure removing this completely is wise, so this will need to be researched.

(In an application the size of WordPress.org, for example, MySQL servers will go away occasionally, and failover as needed.)

**Comment by kkmuffme** [comment]
@JJJ we tested it with millions of pageviews & transactions; I created this code almost a year ago already shortly after I opened the ticket.

It's true that it may go away sometimes, but as I already described on the issue this is irrelevant. WP normally has a timeout value of less than 99 seconds. Even if the connection goes away in this time, the worst that happens is that this request has an error. 
Any following request will already not connect to that db server if the error persists, so there is no issue there.

The cost of the heartbeat are so immense in terms of performance, that even when your db server fails once a day, it's not worth it to check, as only the requests that are active during the exact moment of failure are affected.

**Comment by kkmuffme** [comment]
I now inspected the code again and found some improvements. WIll let you know once it's done

**Comment by kkmuffme** [comment]
Just took a look again, and actually it's fine. The thing I just wanted to check is that by default WP checks AFTER the _do_query if mysql error == 2006 (= gone away), it pings and retries the query.
This is something we may want to add - I just committed it now but it's not tested (bc I don't know how to produce a 2006 on purpose) - so the last commit can be reverted if we don't want to do it like this. Everything in the first 2 commits is production tested though

File: ludicrousdb/includes/class-ludicrousdb.php [hunk 1]
```diff
@@ -1151,16 +1141,28 @@ public function flush() {
...
-				if ( mysql_ping( $dbh ) ) {
-					return true;
+		// we only want to check the connection once per page load, to avoid mysqli_ping slowing down things
+		// this means if the read/write server goes down while the page is loaded, we will have an mysql error instead of graceful failover
+		// however this will happen so rarely & isn't an issue (bc user can just refresh the page) that it's not worth to test/mysqli_ping for every read/write
+		$check_connection = wp_cache_get( $dbh->thread_id, 'non_persistent' );
+		// if the connection was not checked or not established, try to check the connection
+		if( $check_connection === false ) {
+			if ( $this->dbh_type_check( $dbh ) ) {
+				if ( true === $this->use_mysqli ) {
```
Comment (thefrosty) [inline ludicrousdb/includes/class-ludicrousdb.php]: I would suggest keeping your conditionals the same, either use yoda conditions or don't.

Comment (kkmuffme) [reply ludicrousdb/includes/class-ludicrousdb.php]: It happened bc I copied the part from WP core, which used to use yoda (and not all locations are changed yet)
Thanks though

**Comment by JJJ** [comment]
Closing this, in favor of #88.

Please ping me if that pull request does not address this issue satisfactorily.

**Comment by kkmuffme** [comment]
@JJJ ok I get the idea behind the other one and it makes more sense than this pull request actually, since it eliminates the whole ping (including the first one) and is more elegant.

However:
1) when disabling heartbeat & there are mysql errors, these error queries will get written to object cache & the page to page cache (if either enabled).

=> see in my pull request the DONOTCACHEPAGE/wp_suspend_cache_addition

Alternative (might be even better with the new solution):
if we get an error in _do_query we call check_connection() (even if heartbeats are disabled), so it will die if db not connected, otherwise try the query again?


2) it isn't trying to reconnect on error 2006 when heartbeats are disabled (as this is the use case where a mysql ping would make sense)

=> see in my pull request

3) it still has heartbeat pings enabled by default, however they don't make sense for 99% of WordPress use cases and just add server load & overhead/lag since it has to send all the pings

**Comment by kkmuffme** [comment]
@JJJ ??

**Pull request closed without merging**